packages/vite/src/node/server/environment.ts

```diff
@@ -0,0 +1,143 @@
+import type { FetchResult } from 'vite/module-runner'
+import colors from 'picocolors'
+import { Environment } from '../environment'
+import type { ViteDevServer } from '../server'
+import { ERR_OUTDATED_OPTIMIZED_DEP } from '../plugins/optimizedDeps'
+import type { EnvironmentOptions, ResolvedEnvironmentOptions } from '../config'
+import { getDefaultResolvedEnvironmentOptions } from '../config'
+import { mergeConfig } from '../utils'
+import type { FetchModuleOptions } from '../ssr/fetchModule'
+import { fetchModule } from '../ssr/fetchModule'
+import { EnvironmentModuleGraph } from './moduleGraph'
+import type { HMRChannel } from './hmr'
+import { createNoopHMRChannel, getShortName, updateModules } from './hmr'
+import { transformRequest } from './transformRequest'
+import type { TransformResult } from './transformRequest'
+import { ERR_CLOSED_SERVER } from './pluginContainer'
+import type { RemoteEnvironmentTransport } from './environmentTransport'
+
+export interface DevEnvironmentSetup {
+  hot?: false | HMRChannel
+  options?: EnvironmentOptions
+  runner?: FetchModuleOptions & {
+    transport?: RemoteEnvironmentTransport
+  }
+}
+
+// Maybe we will rename this to DevEnvironment
+export class DevEnvironment extends Environment {
+  mode = 'dev' as const // TODO: should this be 'serve'?
+  moduleGraph: EnvironmentModuleGraph
+  server: ViteDevServer
+  /**
+   * @internal
+   */
+  _ssrRunnerOptions: FetchModuleOptions | undefined
+  /**
+   * HMR channel for this environment. If not provided or disabled,
+   * it will be a noop channel that does nothing.
+   *
+   * @example
+   * environment.hot.send({ type: 'full-reload' })
+   */
+  hot: HMRChannel
+  constructor(
+    server: ViteDevServer,
+    name: string,
+    setup?: {
+      hot?: false | HMRChannel
+      options?: EnvironmentOptions
+      runner?: FetchModuleOptions & {
+        transport?: RemoteEnvironmentTransport
+      }
+    },
+  ) {
+    let options =
+      server.config.environments[name] ??
+      getDefaultResolvedEnvironmentOptions(server.config)
+    if (setup?.options) {
+      options = mergeConfig(
+        options,
+        setup?.options,
+      ) as ResolvedEnvironmentOptions
+    }
+    super(name, server.config, options)
```

Comment:
It looks like `DevEnvironment.config` and `BuildEnvironment.config` have somewhat different meanings.
`DevEnvironment.config` is a root config of Vite server, but `BuildEnvironment.config` is the one resolved for each environment with `EnvironmentOptions` merged with root:

https://github.com/vitejs/vite/blob/485f6ac15bcd4d28da9ca6b39d3333ea7bc21579/packages/vite/src/node/build.ts#L1445-L1446

Wouldn't this make accessing the base class's `Environment.config` inconsistent during dev and build?

Reply:
If this is happening, there is a bug. The `environmentOptions` should only be merged with `environment.options.name`, at root it should also have the defaults.

I'm actually thinking that we should deprecate in the `ResolvedConfig` all the environment options at root. This should only be available at `Config` to set the defaults and then disappear so we force access through the proper environment. Accessing the default after it has been resolved should be considered a bug